metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/configuration/metron-env.xml

```diff
@@ -121,6 +121,14 @@
             <empty-value-valid>true</empty-value-valid>
         </value-attributes>
     </property>
+    <property>
```

Comment:
Why do we have a property for the HDFS url?  Shouldn't we get this property from Ambari similar to how we get Zookeeper and Kafka urls?

Reply:
It took me a bit of effort to figure out how we were doing that, actually. The way we grab the existing properties for Zookeeper and Kafka is via the service_advisor code. See:
* https://github.com/apache/metron/blob/96a66c88e7a62867c6287106a6646d198e2e6a81/metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/service_advisor.py#L130
* https://github.com/apache/metron/blob/96a66c88e7a62867c6287106a6646d198e2e6a81/metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/service_advisor.py#L142

You effectively grab the values at cluster install time and provide those as default recommended values for properties that we've defined in the MPack. Here is the path through that maze:
* https://github.com/apache/metron/blob/96a66c88e7a62867c6287106a6646d198e2e6a81/metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/configuration/metron-env.xml#L133
* https://github.com/apache/metron/blob/96a66c88e7a62867c6287106a6646d198e2e6a81/metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/status_params.py#L111
* https://github.com/apache/metron/blob/96a66c88e7a62867c6287106a6646d198e2e6a81/metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py#L157
* https://github.com/apache/metron/blob/96a66c88e7a62867c6287106a6646d198e2e6a81/metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/service_advisor.py#L144

Defined in the env file, loaded into status_params first (so the status commands can use it), inherited by the params os-specific file (for use during install), modified during install by the service_advisor script.

Hopefully it makes sense why I did this (though not necessarily why this is all so esoteric)

Reply:
I did not realize you were populating that in service_advisor.py.  The zookeeper and kafka urls are actually populated in params_linux.py [here](https://github.com/apache/metron/blob/master/metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py#L129) and [here](https://github.com/apache/metron/blob/master/metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py#L129).  I'm not sure which approach is better or what the tradeoffs are.

Reply:
Hm, that's a good find - I did not know that existed. We also seem to be grabbing Zookeeper from [here](https://github.com/apache/metron/blob/master/metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/service_advisor.py#L162) in the service_advisor.py for Solr. But that's apparently independent from our other zk url. Let me see if there's a way to grab HDFS in the same way. I prefer the approach you found.

Reply:
@merrimanr - I played around with this a bit. The one benefit to the way I have it is that it allows the user to change the HDFS url if there's an issue with it for any reason - it will pull the default recommended value from the property `fs.defaultFS` on install. My main concern with this property at all is namenode HA. I can't exactly use the same approach as was done for storm and zookeeper because I can't get the namenode port, i.e.  `hdfs_url = default("/clusterHostInfo/namenode_host", [])` only gives me a host, no port, and there's no other property in clusterHostInfo to obtain that. I was, however, able to find a modification to the service_advisor style that allowed me to grab it from inside the params files, `config["configurations"]["core-site"]["fs.defaultFS"]`. That gives me the full url `hdfs://node1:8020` as desired. The shortcoming here is that there is no way to change this whatsoever without modifying the python files if there's an issue. But I believe this should still work with namenode HA just fine, per [https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-hdfs/HDFSHighAvailabilityWithNFS.html](https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-hdfs/HDFSHighAvailabilityWithNFS.html)

> fs.defaultFS - the default path prefix used by the Hadoop FS client when none is given
> 
> Optionally, you may now configure the default path for Hadoop clients to use the new HA-enabled logical URI. If you used “mycluster” as the nameservice ID earlier, this will be the value of the authority portion of all of your HDFS paths. This may be configured like so, in your core-site.xml file:

```
<property>
  <name>fs.defaultFS</name>
  <value>hdfs://mycluster</value>
</property>
```

I guess I don't have a strong opinion either way - both can work. Exposing the property means that the user can change it if they need, but they will also need to manually manage it in cases where the namenode nameservice ID changes. Not exposing it means they cannot change it at all, however it will always pull the latest ID from any changes with namenode HA, etc. Which do you think is better? In light of the HA probably working fine with fs.defaultFS, I'm leaning towards not exposing the property.

Reply:
I agree, I don't think we should expose the property.  I think we should go with an approach where Ambari can consistently and reliably populate that setting with or without namenode HA enabled.  As long as that condition is satisfied, I'm good with how you choose to do it.

Reply:
@merrimanr - thanks for talking through this - I decided get rid of it in favor of the backend approach. Testing now and getting the test script together for you to start looking at as well.